src/core/worker.ts

```diff
@@ -422,13 +422,30 @@ export class Worker extends EventEmitter {
       this.stringQueues()
     );
 
-    const encodedJob: string = await this.connection.redis["popAndStoreJob"](
-      queueKey,
-      workerKey,
-      new Date().toString(),
-      this.queue,
-      this.name
-    );
+    let encodedJob: string;
+
+    if (this.connection.redis["popAndStoreJob"]) {
+      encodedJob = await this.connection.redis["popAndStoreJob"](
+        queueKey,
+        workerKey,
+        new Date().toString(),
+        this.queue,
+        this.name
+      );
+    } else {
+      encodedJob = await this.connection.redis.lpop(queueKey);
+      if (encodedJob) {
+        await this.connection.redis.set(
+          workerKey,
+          JSON.stringify({
+            run_at: new Date().toString(),
+            queue: this.queue,
+            worker: this.name,
+            payload: JSON.parse(encodedJob),
+          })
+        );
+      }
+    }
```

Comment:
this integration of the new adapter would be a lot cleaner if you implemented interfaces and pick one adapter runtime, instead of if-then-else branches inlined to code.

- RedisClientInterface
- IORedisLuaClient
- IORedisClient
- IORedisMockClient
- ...

Reply:
I'd love to see an example of this this!  I still want to enable folks to defined the connection with just hash of options (port, host, etc), so how would you suggest accomplishing this?

In this case specifically, there's no distinction between the type of connection, just if the custom method is defined.  If we get back support for the `redis` package (not `ioredis`), it doesn't handle custom LUA methods in the same way, so I think this will be more globally useful.  Bringing back `redis` is a whole other can of worms, because it's methods are not promises.

Reply:
it's called factory pattern, picks proper adapter based on input parameters.

like you do already now (I've seen the code).

and whether lua works or not, can be made two different adapter too, adapter does not need to be 1:1 match with a npm package. could be something abstract as well (supports lua, in this case)

Reply:
Can you share where you think we use the factory pattern already?  What would it look like for someone instantiating a new instance of say a Queue?

Reply:
Public interfaces don't have to change, instantiating Queue can stay the same.

I've stared something:
- https://github.com/actionhero/node-resque/pull/477

but ran immediately to questions when refactoring:
- https://github.com/actionhero/node-resque/pull/453/files#r524945800

I'll keep the rest of the question in PR itself.